### PR TITLE
Add fixed-point `WithinTolerance` method and greater-than operators

### PR DESCRIPTION
**Pull request by ffranr**
This pull request introduces the following changes to the `rfqmath` library:

1. **FixedPoint[T].WithinTolerance method**  
   Adds the `WithinTolerance` method to the `FixedPoint[T]` type, allowing two fixed-point numbers to be checked for equality within a specified tolerance in Parts Per Million (PPM).

2. **Greater-than operators for integer types**  
   Adds support for handling the `>` and `>=` comparisons for integer types.

These changes are necessary towards closing https://github.com/lightninglabs/taproot-assets/issues/871.

### CODE REVIEW

File: rfqmath/fixed_point.go [hunk 1]
```diff
@@ -103,6 +103,53 @@ func (f FixedPoint[T]) Equals(other FixedPoint[T]) bool {
...
+
+	// Determine the larger scale between the two fixed-point numbers.
+	// Both values will be scaled to this larger scale to ensure a
+	// consistent comparison.
+	var largerScale uint8
+	if f.Scale > other.Scale {
+		largerScale = f.Scale
+	} else {
+		largerScale = other.Scale
+	}
```
Comment (dstadulis) [inline rfqmath/fixed_point.go:119]: Ignorant comment: Would the case where, if even possible, the FixedPoints are equal, need to be handled? or adding a general error if `largerScale` isn't set

File: rfqmath/fixed_point_test.go [hunk 2]
```diff
@@ -159,6 +159,144 @@ func testFromUint64[N Int[N]](t *rapid.T) {
...
+			tolerancePpm: 100000, // 10% tolerance in PPM
+			withinBounds: true,
+		},
+		{
+			// Case where both prices are zero, should be within
+			// bounds.
+			firstFp:      FixedPointFromUint64[N](0, 0),
+			secondFp:     FixedPointFromUint64[N](0, 0),
+			tolerancePpm: 100000, // any tolerance in PPM
+			withinBounds: true,
```
Comment (dstadulis) [inline rfqmath/fixed_point_test.go:266]: Maybe this covers the equal case referenced above

File: rfqmath/fixed_point_test.go [hunk 3]
```diff
@@ -159,6 +159,144 @@ func testFromUint64[N Int[N]](t *rapid.T) {
...
+		withinBounds bool
+	}
+
+	testCases := []testCase{
+		{
+			// Case where secondFp is 10% less than firstFp,
+			// tolerance allows 11.11% (111111 PPM). Diff within
+			// bounds.
+			firstFp:      FixedPointFromUint64[N](1000000, 1),
+			secondFp:     FixedPointFromUint64[N](900000, 1),
```
Comment (dstadulis) [inline rfqmath/fixed_point_test.go:187]: Are there any overflow cases to handle?
Seems there should be sufficient space:
>log(2**64)/log(1000000*900000) = 1.6

**Comment by dstadulis** [comment]
# Content
PriceOracle will deliver quote
Then internal checks will be done

# fixedPoints
Is a simplification
in the RFQ quotes, the conversion rate between two assets right now there's two different fields, both assets are compared to BTC, which could be a big number
fixedPoints separate the scalar / separate the significant, can transport in a wireMessage without uint64

File: rfqmath/fixed_point.go [hunk 4]
```diff
@@ -103,6 +103,53 @@ func (f FixedPoint[T]) Equals(other FixedPoint[T]) bool {
 	return f.Coefficient.Equals(other.Coefficient) && f.Scale == other.Scale
 }
 
+// WithinTolerance returns true if the two FixedPoint values are within the
+// given tolerance (in parts per million (PPM)).
+func (f FixedPoint[T]) WithinTolerance(
```
Comment (Roasbeef) [inline rfqmath/fixed_point.go:108]: What's the intended use case of this method? To gauge if an action would result in precision loss?

Comment (Roasbeef) [reply rfqmath/fixed_point.go:108]: Looking at the tests, in what case would we ever be manipulating two values with different scale? If you look at the conversion methods, it handles scaling up and then down.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: rfq/negotiator.go [hunk 5]
```diff
@@ -518,6 +518,8 @@ func expiryWithinBounds(expiryUnixTimestamp uint64,
 
 // priceWithinBounds returns true if the difference between the first price and
 // the second price is within the given tolerance (in parts per million (PPM)).
+//
+// TODO(ffranr): Replace with FixedPoint[T].WithinTolerance.
```
Comment (GeorgeTsagk) [inline rfq/negotiator.go:522]: is this meant to be addressed in follow-up PR?

File: rfqmath/fixed_point_test.go [hunk 6]
```diff
@@ -159,6 +159,144 @@ func testFromUint64[N Int[N]](t *rapid.T) {
...
+			secondFp:     FixedPointFromUint64[N](100_000, 4),
+			tolerancePpm: 100000, // any tolerance in PPM
+			withinBounds: false,
+		},
+		{
+			// Case where secondFp is zero and firstFp is
+			// non-zero, should not be within bounds.
+			firstFp:      FixedPointFromUint64[N](100_000, 4),
+			secondFp:     FixedPointFromUint64[N](0, 0),
+			tolerancePpm: 100000, // any tolerance in PPM
```
Comment (GeorgeTsagk) [inline rfqmath/fixed_point_test.go:281]: How would we pick a `tolerancePpm` value in practice?

File: rfqmath/fixed_point_test.go [hunk 7]
```diff
@@ -159,6 +159,144 @@ func testFromUint64[N Int[N]](t *rapid.T) {
...
+			withinBounds: false,
+		},
+	}
+
+	// Run the test cases.
+	for idx, tc := range testCases {
```
Comment (GeorgeTsagk) [inline rfqmath/fixed_point_test.go:287]: would it make sense to add a case where there's a huge scale difference, in order to see precision getting exposed?

i.e
```
firstFp:      FixedPointFromUint64[N](314, 2),
secondFp:     FixedPointFromUint64[N](314_159_265_359, 11),
```

File: rfqmath/fixed_point_test.go (hunk 6 above)
Comment (GeorgeTsagk) [reply rfqmath/fixed_point_test.go:281]: is this value configured by the user, to express a variable exposure to volatility?

File: rfq/negotiator.go (hunk 5 above)
Comment (ffranr) [reply rfq/negotiator.go:522]: Yes.

File: rfqmath/fixed_point_test.go (hunk 6 above)
Comment (ffranr) [reply rfqmath/fixed_point_test.go:281]: We currently use:
```
	AcceptPriceDeviationPpm uint64 `long:"acceptpricedeviationppm" description:"The default price deviation in parts per million that is accepted by the RFQ negotiator"`
```

as the tolerance. We will use that same config field with this new functionality.

File: rfqmath/fixed_point.go (hunk 4 above)
Comment (ffranr) [reply rfqmath/fixed_point.go:108]: > What's the intended use case of this method? To gauge if an action would result in precision loss?

This TODO describes one use case:
https://github.com/lightninglabs/taproot-assets/pull/1135/files#diff-64a27caf98eee653f23ffa833568a598241835b52d96fa4e26fc4127f92ad4f3R522

The current code uses this function:
```
pricesWithinBounds(firstPrice lnwire.MilliSatoshi,
	secondPrice lnwire.MilliSatoshi, tolerancePpm uint64) bool
```

I think we will need to replace that function with one that supports fixed-points.

Comment (ffranr) [reply rfqmath/fixed_point.go:108]: > Looking at the tests, in what case would we ever be manipulating two values with different scale? If you look at the conversion methods, it handles scaling up and then down.

`WithinTolerance` can handle fixed-point numbers of different scales. This feature may not always be necessary, but it's implemented to simplify usage, removing the need to scale arguments before calling the function.

We may need to handle fixed-point values at different scales in `pricesWithinBounds` callsites, depending on the values provided by our RFQ peer or the price oracle.

File: rfqmath/fixed_point_test.go (hunk 3 above)
Comment (ffranr) [reply rfqmath/fixed_point_test.go:187]: `WithinTolerance` uses methods from the `rfqmath.Arithmetic` interface for calculations, so I think the overflow responsibility should be on those methods. I'm not sure if its possible to get the `Mul` calls to overflow.

File: rfqmath/fixed_point.go (hunk 1 above)
Comment (ffranr) [reply rfqmath/fixed_point.go:119]: I've added a test case where both fixed-points are exactly equal:
```
		{
			// Case where firstFp and secondFp are equal,
			// tolerance is 0 PPM. Diff within bounds.
			firstFp:      FixedPointFromUint64[N](100_000, 2),
			secondFp:     FixedPointFromUint64[N](100_000, 2),
			tolerancePpm: 0, // 0% tolerance in PPM
			withinBounds: true,
		},
```

**Comment by ffranr** [comment]
Use case for this new method is replacing `pricesWithinBounds`, see:
https://github.com/lightninglabs/taproot-assets/pull/1135/files#diff-64a27caf98eee653f23ffa833568a598241835b52d96fa4e26fc4127f92ad4f3R522

File: rfqmath/fixed_point_test.go [hunk 8]
```diff
@@ -159,6 +159,169 @@ func testFromUint64[N Int[N]](t *rapid.T) {
...
+			secondFp:     FixedPointFromUint64[N](0, 0),
+			tolerancePpm: 100000, // any tolerance in PPM
+			withinBounds: true,
+		},
+		{
+			// Case where firstFp is zero and secondFp is
+			// non-zero, should not be within bounds.
+			firstFp:      FixedPointFromUint64[N](0, 0),
+			secondFp:     FixedPointFromUint64[N](100_000, 4),
+			tolerancePpm: 100000, // any tolerance in PPM
```
Comment (gijswijs) [inline rfqmath/fixed_point_test.go:298]: Any? 1_000_000 should be within bounds, right? Same for the case below.

**Review by gijswijs — COMMENTED** [review]
Nice looking PR. Made some comments throughout. 

I'm still not fully understanding the use case tho:

This will be used during the handling of sell accept messages and buy accept messages that the node receives from its peer. The bid price or the ask price in those messages is compared to the bid price or the ask price respectively that our oracle provides. If it's within bounds (specified in ppm) we accept the offer.

But iiuc at that point we are already comparing prices converted to BTC?  But the `WithinTolerance` works with `FixedPoint[T]` which I very much associate with taproot assets.

Shouldn't `WithinTolerance` work with `lnwire.MilliSatoshi` parameters?

**Comment by ffranr** [comment]
> Nice looking PR. Made some comments throughout.
> 
> I'm still not fully understanding the use case tho:
> 
> This will be used during the handling of sell accept messages and buy accept messages that the node receives from its peer. The bid price or the ask price in those messages is compared to the bid price or the ask price respectively that our oracle provides. If it's within bounds (specified in ppm) we accept the offer.
> 
> But iiuc at that point we are already comparing prices converted to BTC? But the `WithinTolerance` works with `FixedPoint[T]` which I very much associate with taproot assets.
> 
> Shouldn't `WithinTolerance` work with `lnwire.MilliSatoshi` parameters?

This PR is in preparation for a follow-up PR. In the follow-up, the price oracle and RFQ wire messages will communicate asset-to-BTC rates instead of  BTC prices. The current setup is a workaround where the rate coefficient is effectively typed as a millisatoshi amount. The follow-up PR will address and correct this.

Basically we want to compare rates and not prices. And the "prices" we're currently dealing with are rates in disguise.

File: rfqmath/fixed_point_test.go [hunk 9]
```diff
@@ -175,4 +338,6 @@ func TestFixedPoint(t *testing.T) {
 	t.Run("equality", rapid.MakeCheck(testEquality[BigInt]))
 
 	t.Run("from_uint64", rapid.MakeCheck(testFromUint64[BigInt]))
+
+	t.Run("within_tolerance", rapid.MakeCheck(testWithinTolerance[BigInt]))
```
[the remark this answers is not in the document]
Comment (ffranr) [reply rfqmath/fixed_point_test.go:342]: Just following the existing pattern. I might add random data inputs if I can think of something useful.

File: rfqmath/fixed_point_test.go (hunk 8 above)
Comment (ffranr) [reply rfqmath/fixed_point_test.go:298]: Yes, I've revised the comment to say: tolerance not effectual

File: rfqmath/fixed_point_test.go (hunk 3 above)
Comment (Roasbeef) [reply rfqmath/fixed_point_test.go:187]: > I'm not sure if its possible to get the Mul calls to overflow.

So we'll use the `BigInt[T]` version everywhere, which for all practical purposes, can't actually overflow.

File: rfqmath/fixed_point_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply rfqmath/fixed_point_test.go:342]: If you want to use `rapid`, then the tests need to be written in a more black-box style. So asserting some invariants/properties based on the desired behavior of the function in the abstract. 

Non-blocking though.

[the remark this answers is not in the document]
Comment (Roasbeef) [reply rfqmath/fixed_point_test.go:342]: I think we can retain this current TDD test (moving it outside the scope of this function), and also add this property based test to supplement things: 
```go
package rfqmath

import (
	"math/big"
	"testing"

	"github.com/stretchr/testify/require"
	"pgregory.net/rapid"
)

func TestWithinToleranceBigInt(t *testing.T) {
	rapid.Check(t, func(t *rapid.T) {
		// Create a require instance
		req := require.New(t)

		// Generate random BigInts and scales
		a := rapid.Uint64().Draw(t, "a")
		b := rapid.Uint64().Draw(t, "b")
		scaleA := rapid.Uint8Range(0, 18).Draw(t, "scaleA")
		scaleB := rapid.Uint8Range(0, 18).Draw(t, "scaleB")
		tolerancePpm := rapid.Uint64Range(0, 1000000).Draw(t, "tolerancePpm")

		fpA := FixedPointFromUint64[BigInt](a, scaleA)
		fpB := FixedPointFromUint64[BigInt](b, scaleB)

		// Test property: WithinTolerance should be reflexive
		req.True(fpA.WithinTolerance(fpA, NewBigInt(big.NewInt(int64(tolerancePpm)))),
			"A number should be within tolerance of itself")

		// Test property: WithinTolerance should be symmetric
		resultAB := fpA.WithinTolerance(fpB, NewBigInt(big.NewInt(int64(tolerancePpm))))
		resultBA := fpB.WithinTolerance(fpA, NewBigInt(big.NewInt(int64(tolerancePpm))))
		req.Equal(resultAB, resultBA,
			"WithinTolerance should be symmetric")

		// Test property: Zero tolerance should only allow exact equality
		zeroTolerance := NewBigInt(big.NewInt(0))
		req.Equal(fpA.WithinTolerance(fpB, zeroTolerance), fpA.Equals(fpB),
			"Zero tolerance should only allow exact equality")

		// Test property: Maximum tolerance (1,000,000 PPM) should always return true
		maxTolerance := NewBigInt(big.NewInt(1000000))
		req.True(fpA.WithinTolerance(fpB, maxTolerance),
			"Maximum tolerance should always return true")

		// Test property: Tolerance behavior near boundaries
		if !fpA.Equals(fpB) {
			largerScale := scaleA
			if scaleB > scaleA {
				largerScale = scaleB
			}
			fpAScaled := fpA.ScaleTo(largerScale)
			fpBScaled := fpB.ScaleTo(largerScale)

			var delta *big.Int
			if fpAScaled.Coefficient.Gt(fpBScaled.Coefficient) {
				delta = new(big.Int).Sub(fpAScaled.Coefficient.value, fpBScaled.Coefficient.value)
			} else {
				delta = new(big.Int).Sub(fpBScaled.Coefficient.value, fpAScaled.Coefficient.value)
			}

			maxCoeff := new(big.Int).Set(fpAScaled.Coefficient.value)
			if fpBScaled.Coefficient.Gt(fpAScaled.Coefficient) {
				maxCoeff.Set(fpBScaled.Coefficient.value)
			}

			// Calculate the exact PPM difference
			ppmDiff := new(big.Int).Mul(delta, big.NewInt(1000000))
			ppmDiff.Div(ppmDiff, maxCoeff)

			// Test that WithinTolerance returns true for tolerances greater than or equal to ppmDiff
			req.True(fpA.WithinTolerance(fpB, NewBigInt(new(big.Int).Add(ppmDiff, big.NewInt(1)))),
				"Should be within tolerance for tolerance > ppmDiff")

			// Test that WithinTolerance returns false for tolerances less than ppmDiff
			if ppmDiff.Sign() > 0 {
				req.False(fpA.WithinTolerance(fpB, NewBigInt(new(big.Int).Sub(ppmDiff, big.NewInt(1)))),
					"Should not be within tolerance for tolerance < ppmDiff")
			}
		}
	})
}
```

[the remark this answers is not in the document]
Comment (Roasbeef) [reply rfqmath/fixed_point_test.go:342]: A variant that breaks things up a bit, and uses `float64` to compute the final tolerance values directly: 
```
package rfqmath_test

import (
    "math"
    "math/big"
    "testing"

    "pgregory.net/rapid"

    "github.com/yourusername/yourmodule/rfqmath"
)

func TestWithinTolerance_EqualValues(t *testing.T) {
    rapid.Check(t, func(t *rapid.T) {
        // Generate a random coefficient and scale
        coef := t.Int64()
        scale := uint8(t.Uint32Range(0, 18))
        tolerancePpm := t.Uint64()

        // Create two identical FixedPoint[BigInt] values
        f1 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef)),
            Scale:       scale,
        }

        f2 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef)),
            Scale:       scale,
        }

        tolerancePpmBigInt := rfqmath.NewBigInt(big.NewInt(int64(tolerancePpm)))

        // The result should always be true when the values are equal
        result := f1.WithinTolerance(f2, tolerancePpmBigInt)
        if !result {
            t.Fatalf("WithinTolerance should be true when values are equal, but got false")
        }
    })
}

func TestWithinTolerance_ToleranceZero(t *testing.T) {
    rapid.Check(t, func(t *rapid.T) {
        // Generate random coefficients and scales
        coef1 := t.Int64()
        coef2 := t.Int64()
        scale1 := uint8(t.Uint32Range(0, 18))
        scale2 := uint8(t.Uint32Range(0, 18))

        f1 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef1)),
            Scale:       scale1,
        }

        f2 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef2)),
            Scale:       scale2,
        }

        tolerancePpmBigInt := rfqmath.NewBigInt(big.NewInt(0))

        result := f1.WithinTolerance(f2, tolerancePpmBigInt)

        // Scale both to the larger scale for comparison
        largerScale := scale1
        if scale2 > scale1 {
            largerScale = scale2
        }
        scaledF1 := f1.ScaleTo(largerScale)
        scaledF2 := f2.ScaleTo(largerScale)

        equal := scaledF1.Coefficient.Equals(scaledF2.Coefficient)

        if result != equal {
            t.Fatalf("WithinTolerance with zero tolerance mismatch: expected %v, got %v", equal, result)
        }
    })
}

func TestWithinTolerance_Symmetric(t *testing.T) {
    rapid.Check(t, func(t *rapid.T) {
        // Generate random coefficients, scales, and tolerance
        coef1 := t.Int64()
        coef2 := t.Int64()
        scale1 := uint8(t.Uint32Range(0, 18))
        scale2 := uint8(t.Uint32Range(0, 18))
        tolerancePpm := t.Uint64()

        f1 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef1)),
            Scale:       scale1,
        }

        f2 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef2)),
            Scale:       scale2,
        }

        tolerancePpmBigInt := rfqmath.NewBigInt(big.NewInt(int64(tolerancePpm)))

        result1 := f1.WithinTolerance(f2, tolerancePpmBigInt)
        result2 := f2.WithinTolerance(f1, tolerancePpmBigInt)

        if result1 != result2 {
            t.Fatalf("WithinTolerance is not symmetric: f1.WithinTolerance(f2)=%v, f2.WithinTolerance(f1)=%v", result1, result2)
        }
    })
}

func TestWithinTolerance_MaxTolerance(t *testing.T) {
    rapid.Check(t, func(t *rapid.T) {
        // Generate random coefficients and scales
        coef1 := t.Int64()
        coef2 := t.Int64()
        scale1 := uint8(t.Uint32Range(0, 18))
        scale2 := uint8(t.Uint32Range(0, 18))
        // Set tolerancePpm to a large value
        tolerancePpm := t.Uint64Range(1_000_000, 10_000_000_000)

        f1 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef1)),
            Scale:       scale1,
        }

        f2 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef2)),
            Scale:       scale2,
        }

        tolerancePpmBigInt := rfqmath.NewBigInt(big.NewInt(int64(tolerancePpm)))

        // The result should always be true when tolerancePpm is large
        result := f1.WithinTolerance(f2, tolerancePpmBigInt)
        if !result {
            t.Fatalf("WithinTolerance should be true when tolerancePpm is large, but got false")
        }
    })
}

func TestWithinTolerance_Property(t *testing.T) {
    rapid.Check(t, func(t *rapid.T) {
        // Generate random coefficients within a reasonable range
        coef1 := t.Int64Range(-1_000_000_000, 1_000_000_000)
        coef2 := t.Int64Range(-1_000_000_000, 1_000_000_000)

        scale1 := uint8(t.Uint32Range(0, 9))
        scale2 := uint8(t.Uint32Range(0, 9))

        tolerancePpm := t.Uint64Range(0, 1_000_000)

        f1 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef1)),
            Scale:       scale1,
        }

        f2 := rfqmath.FixedPoint[rfqmath.BigInt]{
            Coefficient: rfqmath.NewBigInt(big.NewInt(coef2)),
            Scale:       scale2,
        }

        tolerancePpmBigInt := rfqmath.NewBigInt(big.NewInt(int64(tolerancePpm)))

        result := f1.WithinTolerance(f2, tolerancePpmBigInt)

        // Compute expected result using float64
        f1Float := f1.ToFloat64()
        f2Float := f2.ToFloat64()

        delta := math.Abs(f1Float - f2Float)
        maxVal := math.Max(math.Abs(f1Float), math.Abs(f2Float))

        tolerance := (float64(tolerancePpm) / 1_000_000) * maxVal

        expected := delta <= tolerance

        if result != expected {
            t.Fatalf("WithinTolerance mismatch:\n"+
                "f1 = %v (float: %f),\n"+
                "f2 = %v (float: %f),\n"+
                "tolerancePpm = %v,\n"+
                "delta = %e,\n"+
                "tolerance = %e,\n"+
                "result = %v,\n"+
                "expected = %v",
                f1, f1Float, f2, f2Float, tolerancePpm, delta, tolerance, result, expected)
        }
    })
}
```

**Review by Roasbeef — APPROVED** [review]
LGTM 🎽

**Comment by ffranr** [comment]
I've added the property based tests that roas suggested.

File: rfqmath/fixed_point_test.go [hunk 10]
```diff
@@ -159,6 +160,375 @@ func testFromUint64[N Int[N]](t *rapid.T) {
...
+		{
+			// Case where secondFp is greater than firstFp,
+			// tolerance allows 5% (50000 PPM). Diff within bounds.
+			firstFp: FixedPoint[N]{
+				Coefficient: NewInt[N]().FromUint64(314),
+				Scale:       2,
+			},
+			secondFp: FixedPoint[N]{
+				Coefficient: NewInt[N]().FromUint64(
+					314_159_265_359,
```
Comment (GeorgeTsagk) [inline rfqmath/fixed_point_test.go:261]: :+1:

**Review by GeorgeTsagk — APPROVED** [review]
LGTM

**Pull request merged**